Expose network mode in the DockerBuildImage task

tlusk opened the pull request:
This simply exposes the docker-java property so it can be used in the Gradle task.

This cannot be merged in until the corresponding pull request in docker-java has been merged and a new release has been made.

See: https://github.com/docker-java/docker-java/pull/1061

closes #555

cdancy commented:
@tlusk thanks for the PR! If and when this gets merged with those guys, and they spin a new release, feel free to ping us on this end and we will go ahead and merge this PR.

tlusk commented:
@cdancy The upstream changes for this have been released in docker-java 3.1.0 RC4

bmuschko commented:
@tlusk I don't think we'll want to pull a release candidate version of Docker Java and release with it. We are going to wait until the final is out.

cdancy commented:
@bmuschko while I agree with the sentiment we're using an RC version now. Those guys haven't done a proper release since last September and who knows when the next one will be done. The maintainers have said that master, and thus the RC releases, are stable and not failing and testing with them does suggest this is true.

It's for this reason and handful of others I've floated around the idea of using spotify's docker-client lib which at the time was still getting off the ground but has now surpassed feature-wise the docker-java library. It would take some work to switch but might be worth it as it seems to be better maintained and more mature.

bmuschko commented:
> It's for this reason and handful of others I've floated around the idea of using spotify's docker-client lib which at the time was still getting off the ground but has now surpassed feature-wise the docker-java library. It would take some work to switch but might be worth it as it seems to be better maintained and more mature.

I am all in for it! I was actually looking into switching over years ago. It'd be worth putting together a prototype. I might look into it but want to fix a couple of other issues beforehand.

cdancy commented:
@bmuschko that would be great and would probably be a fair amount of work so we can divide things up when the time comes.

cdancy commented:
They also have an uber/shadow/fat jar that they publish and would remove us (myself mainly) from having to host our own shaded version of `docker-java` (to get around gradles classpath cloberring) because those guys don't have the time to code things.

tanelso2 commented:
Is there any chance of this getting merged soon? This would be very useful for our use case

tlusk commented:
@cdancy I noticed you bumped up the `docker-java` version to `rc5` so this MR should be good to merge in now.

I rebased and added a basic functional test case for this flag.

cdancy commented:
@tlusk taking it in now. Thanks for the PR!
